plugins/database/mongodb/mongodb.go

```diff
@@ -119,83 +149,146 @@ func (m *MongoDB) CreateUser(ctx context.Context, statements dbplugin.Statements
+	resp := newdbplugin.NewUserResponse{
+		Username: username,
+	}
+	return resp, nil
 }
 
-// SetCredentials uses provided information to set/create a user in the
-// database. Unlike CreateUser, this method requires a username be provided and
-// uses the name given, instead of generating a name. This is used for creating
-// and setting the password of static accounts, as well as rolling back
-// passwords in the database in the event an updated database fails to save in
-// Vault's storage.
-func (m *MongoDB) SetCredentials(ctx context.Context, statements dbplugin.Statements, staticUser dbplugin.StaticUserConfig) (username, password string, err error) {
-	// Grab the lock
+func newUsername(config newdbplugin.UsernameMetadata) (string, error) {
+	displayName := trunc(config.DisplayName, 15)
+	roleName := trunc(config.RoleName, 15)
+
+	userUUID, err := credsutil.RandomAlphaNumeric(20, false)
+	if err != nil {
+		return "", err
+	}
+
+	now := fmt.Sprint(time.Now().Unix())
+
+	parts := []string{
+		"v",
+		displayName,
+		roleName,
+		userUUID,
+		now,
+	}
+	username := joinNonEmpty("-", parts...)
+	username = trunc(username, 100)
+
+	return username, nil
+}
+
+func trunc(str string, l int) string {
+	if len(str) < l {
+		return str
+	}
+	return str[:l]
+}
+
+func joinNonEmpty(sep string, vals ...string) string {
+	if sep == "" {
+		return strings.Join(vals, sep)
+	}
+	switch len(vals) {
+	case 0:
+		return ""
+	case 1:
+		return vals[0]
+	}
+	builder := &strings.Builder{}
+	for _, val := range vals {
+		if val == "" {
+			continue
+		}
+		if builder.Len() > 0 {
+			builder.WriteString(sep)
+		}
+		builder.WriteString(val)
+	}
+	return builder.String()
+}
+
+func removeEmpty(input []string) []string {
+	output := []string{}
+	for _, val := range input {
+		if val != "" {
+			output = append(output, val)
+		}
+	}
+	return output
+}
+
+func (m *MongoDB) UpdateUser(ctx context.Context, req newdbplugin.UpdateUserRequest) (newdbplugin.UpdateUserResponse, error) {
+	if req.Password != nil {
+		err := m.changeUserPassword(ctx, req.Username, req.Password.NewPassword)
+		return newdbplugin.UpdateUserResponse{}, err
+	}
+	return newdbplugin.UpdateUserResponse{}, nil
+}
+
+func (m *MongoDB) changeUserPassword(ctx context.Context, username, password string) error {
 	m.Lock()
 	defer m.Unlock()
 
-	username = staticUser.Username
-	password = staticUser.Password
+	connURL := m.getConnectionURL()
+	cs, err := connstring.Parse(connURL)
+	if err != nil {
+		return err
+	}
 
+	// Currently doesn't support custom statements for changing the user's password
 	changeUserCmd := &updateUserCommand{
 		Username: username,
 		Password: password,
 	}
 
-	connURL := m.getConnectionURL()
-	cs, err := connstring.Parse(connURL)
-	if err != nil {
-		return "", "", err
-	}
-	if err := m.runCommandWithRetry(ctx, cs.Database, changeUserCmd); err != nil {
-		return "", "", err
+	database := cs.Database
+	if username == m.Username || database == "" {
+		database = "admin"
 	}
```

Comment:
The default is there because it isn't always specified in the connection string but is a reasonable default since users are usually created in the `admin` database (I'm not 100% sure if you can create them anywhere else). For what it's worth, this is existing behavior: https://github.com/hashicorp/vault/blob/master/plugins/database/mongodb/mongodb.go#L116-L119